tools/mount-docker-image/main.c

```diff
@@ -0,0 +1,305 @@
+        free(cmd);
+
+    if (is)
+        pclose(is);
+
+    return ret;
+}
+
+typedef struct
+{
+    char* lowerdir;
+    char* upperdir;
+    char* workdir;
+} parse_results_t;
+
+static json_result_t _json_read_callback(
+    json_parser_t* parser,
+    json_reason_t reason,
+    json_type_t type,
+    const json_union_t* un,
+    void* callback)
+{
+    json_result_t result = JSON_UNEXPECTED;
+    parse_results_t* data = (parse_results_t*)callback;
+
+    switch (reason)
+    {
+        case JSON_REASON_NONE:
+        {
+            /* Unreachable */
+            assert(false);
+            break;
+        }
+        case JSON_REASON_NAME:
+        case JSON_REASON_BEGIN_OBJECT:
+        case JSON_REASON_END_OBJECT:
+        case JSON_REASON_BEGIN_ARRAY:
+        case JSON_REASON_END_ARRAY:
+        {
+            /* ignore these */
+            break;
+        }
+        case JSON_REASON_VALUE:
+        {
+            if (json_match(parser, "GraphDriver.Data.LowerDir") == JSON_OK)
+            {
+                if (type != JSON_TYPE_STRING)
+                {
+                    result = JSON_TYPE_MISMATCH;
+                    goto done;
+                }
+
+                if (!(data->lowerdir = strdup(un->string)))
+                    _err("out of memory");
+            }
+            else if (json_match(parser, "GraphDriver.Data.UpperDir") == JSON_OK)
+            {
+                if (type != JSON_TYPE_STRING)
+                {
+                    result = JSON_TYPE_MISMATCH;
+                    goto done;
+                }
+
+                if (!(data->upperdir = strdup(un->string)))
+                    _err("out of memory");
+            }
+            else if (json_match(parser, "GraphDriver.Data.WorkDir") == JSON_OK)
+            {
+                if (type != JSON_TYPE_STRING)
+                {
+                    result = JSON_TYPE_MISMATCH;
+                    goto done;
+                }
+
+                if (!(data->workdir = strdup(un->string)))
+                    _err("out of memory");
+            }
+            else
+            {
+                /* ignore unknown tags */
+            }
+
+            break;
+        }
+    }
+
+    result = JSON_OK;
+
+done:
+    return result;
+}
+
+static void _parse(char* data, size_t size, parse_results_t* results)
+{
+    json_parser_t parser;
+    json_result_t r;
+    static json_allocator_t allocator = {
+        malloc,
+        free,
+    };
+
+    const json_parser_options_t options = {1};
+
+    if ((r = json_parser_init(
+             &parser,
+             data,
+             size,
+             _json_read_callback,
+             results,
+             &allocator,
+             &options)) != JSON_OK)
+    {
+        _err("json_parser_init() failed");
+    }
+
+    if ((r = json_parser_parse(&parser)) != JSON_OK)
+        _err("json_parser_parse() failed");
+
+    if (parser.depth != 0)
+        _err("json parsing failed");
+
+    if (!results->lowerdir)
```

Comment:
For base-images, there does not seem to be a lower-dir. E.g `docker inspect alpine`
In that case, not sure if we can omit the lowerdir option in the mount command or need to specify an empty folder.

Reply:
Good idea about validation of mont-docker-image and export! I will put something together.

Also, I was not aware of the lowerdir problem. I will try to fix it.

Reply:
I compared exported and mounted docker images. It turns out that exported images contains extra files since they were exported from live images or containers (whereas mounted images are from static images). These include:

```
/dev/console
/dev/pts
/dev/shm
.dockerenv
localhost.localdomain
/etc/mtab
```

But I think these should not be in Mystikos images anyway, so this is probably good.

Reply:
A general test that compares will not be feasible since it would included many exclusions and it would have to ignore dead soft links (which there are many of in docker images).